### ISSUE TITLE
Issues with responses from getImage & getImageScaledToCanvas

### ISSUE THREAD

**Issue by oyeanuj** [mosch/react-avatar-editor#225]
Hey @mosch, back with another question :) So, I was trying to generate previews with filters, and the way I was doing that was that once the image was dropped by the user, in `componentDidUpdate` of the component (that has `AvatarEditor`), I use `this.editor.getImageScaledToCanvas().toDataURL()` to generate a Base64 preview that the preview components use.

What I noticed that initially the preview generated was always a blank image (even though the editor is showing an image at that point), but if I changed position/scale/rotation, then the preview was generated correctly.

Any thoughts on what might be happening and possible workarounds?

**Comment by oyeanuj** [comment]
@mosch I'm modifying the original thread to add a couple of related observation/concerns about `getImageScaledToCanvas` and `getImage` functions.

1. (mentioned above): `getImageScaledToCanvas` returns a blank image when image is dropped.

2. If I start with an image already present, then in `componentDidMount`, `getImageScaledToCanvas` returns an image but `getImage` doesn't.

3. I'd like to suggest renaming these two functions to something along the lines of `getCanvasWithOriginalImage` and `getCanvasWithScaledImage` since those names are more accurate. I always find myself getting tricked expecting the image to be returned but instead getting a `canvas` element back.

FWIW, I found my workarounds but still wanted to put these issues out there.

**Comment by mosch** [comment]
@oyeanuj I like what you suggest. I agree that the names are not very good and we can improve them. Let's say we add the two names and mark the others as deprecated.

**Comment by oyeanuj** [comment]
Great, any thoughts on (1) or (2) above - what might be going on there?

**Comment by mosch** [comment]
Hmm, had a quick look in the code. But I could not find any obvious there. Would need some debugging to get an better idea. Could you do more research or do you need help on this?

**Comment by oyeanuj** [comment]
@mosch I did a little bit more digging and seems like they have a common root cause.

TLDR; `image.resource` is being set as a `then` to the promise in `handleImageReady` which is being called after the calls to `getImage` / `getImageScaledToCanvas` in `componentDidMount of user's wrapper. So, essentially, a race condition.

---

Details:

 It seems like within the `this.state.image` object, the `resource` key doesn't exist at the point of calling them right after loading an image (`componentDidMount`). The only thing which exists in the object at that point is `x` and `y` keys.

Given that, `getImage` throws an error here since `resource` is `undefined`: 
https://github.com/mosch/react-avatar-editor/blob/6e21bf9f7aa600b4103f0fb8c405aefe7cb6c51b/src/index.js#L291-L298

For `getImageScaledToCanvas`, it calls `paintImage` which draws something only if `image.resource` is available: 
https://github.com/mosch/react-avatar-editor/blob/6e21bf9f7aa600b4103f0fb8c405aefe7cb6c51b/src/index.js#L452-L453

I noticed that the `resource` is being set in `handleImageReady` which is called as a `then` to the promise here:
https://github.com/mosch/react-avatar-editor/blob/master/src/index.js#L408-L411

And `handleImageReady` is being entered after the call to `getImage`/`getImageScaledToCanvas`. Not sure what is the best way to deal with the race condition but I hope this will help us get closer to the bug.
